test/user_permissions.go

```diff
@@ -0,0 +1,71 @@
+package test
+
+import (
+	"context"
+	"testing"
+
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type AllowedOperations struct {
+	Get    bool
+	List   bool
+	Create bool
+	Delete bool
+}
+
+var (
+	AllowAll = AllowedOperations{
+		Get:    true,
+		List:   true,
+		Create: true,
+		Delete: true,
+	}
+	AllowViewOnly = AllowedOperations{
+		Get:  true,
+		List: true,
+	}
+)
+
+type UserPermissionTest struct {
+	Name              string
+	UserContext       *Context
+	AllowedOperations map[schema.GroupVersionResource]AllowedOperations
+}
+
+func RunUserPermissionTests(t *testing.T, objects map[schema.GroupVersionResource]*unstructured.Unstructured, tests ...UserPermissionTest) {
+	for _, test := range tests {
+		t.Run(test.Name, func(t *testing.T) {
+			for gvr, allowed := range test.AllowedOperations {
+				client := test.UserContext.Clients.Dynamic.Resource(gvr).Namespace("serverless-tests")
+
+				obj := objects[gvr].DeepCopy()
+				obj.SetName("test-" + gvr.Resource)
+
+				_, err := client.Create(context.Background(), obj, metav1.CreateOptions{})
+				if (allowed.Create && err != nil) || (!allowed.Create && !apierrs.IsForbidden(err)) {
+					t.Errorf("Unexpected error creating %s, allowed = %v, err = %v", gvr.String(), allowed.Create, err)
+				}
+
+				err = client.Delete(context.Background(), obj.GetName(), metav1.DeleteOptions{})
+				if (allowed.Delete && err != nil) || (!allowed.Delete && !apierrs.IsForbidden(err)) {
+					t.Errorf("Unexpected error deleting %s, allowed = %v, err = %v", gvr.String(), allowed.Delete, err)
+				}
+
```

Comment:
you will include the `PollImmediate` section here too, right?

Reply:
after the other PR merges, yep

Reply:
cool, thanks

Reply:
Done.